src/shogun/lib/SGVector.cpp

```diff
@@ -77,24 +77,38 @@ SGVector<T>::SGVector() : SGReferencedData()
 
 template<class T>
 SGVector<T>::SGVector(T* v, index_t len, bool ref_counting)
-: SGReferencedData(ref_counting), vector(v), vlen(len)
+: SGReferencedData(ref_counting), vector(v), vlen(len), gpu_vector(NULL)
 {
 }
 
 template<class T>
 SGVector<T>::SGVector(index_t len, bool ref_counting)
-: SGReferencedData(ref_counting), vlen(len)
+: SGReferencedData(ref_counting), vlen(len), gpu_vector(NULL)
 {
 	vector=SG_MALLOC(T, len);
 }
 
 template<class T>
+SGVector<T>::SGVector(GPUMemoryBase<T>* vector, index_t len)
+ : SGReferencedData(true), vector(NULL), vlen(len), gpu_vector(vector)
+{
+}
+
+template<class T>
 SGVector<T>::SGVector(const SGVector &orig) : SGReferencedData(orig)
 {
 	copy_data(orig);
 }
 
 template<class T>
+SGVector<T>& SGVector<T>::operator=(SGVector<T> const &orig)
+{
+	copy_refcount(orig);
+	ref();
+	copy_data(orig);
```

Comment:
data copying implicitly is not really good in my opinion....
I would do a shallow copy on operator=